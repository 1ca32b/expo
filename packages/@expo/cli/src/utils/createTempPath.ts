import { randomBytes } from 'crypto';
import fs from 'fs';
import { tmpdir } from 'os';
import path from 'path';

const uniqueTempPath = (): string => {
  const tempDir = fs.realpathSync(tmpdir());
  const uniqueString = randomBytes(16).toString('hex');
  return path.join(tempDir, uniqueString);
};

// Functionally equivalent to: https://github.com/sindresorhus/tempy/blob/943ade0c935367117adbe2b690516ebc94139c6d/index.js#L43-L47
export function createTempDirectoryPath(): string {
  const directory = uniqueTempPath();
  fs.mkdirSync(directory);
  return directory;
}

// Functionally equivalent to: https://github.com/sindresorhus/tempy/blob/943ade0c935367117adbe2b690516ebc94139c6d/index.js#L25-L39
export function createTempFilePath(name = ''): string {
  if (name) {
    return path.join(createTempDirectoryPath(), name);
  } else {
    return uniqueTempPath();
  }
}
