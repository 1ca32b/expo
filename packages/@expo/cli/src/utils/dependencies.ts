import { getPackageJson } from '@expo/config';

/** @returns true if the expo-dev-client package is found in the project `package.json` file. */
export function hasDirectDevClientDependency(projectRoot: string): boolean {
  const pkg = getPackageJson(projectRoot);
  return !!pkg.dependencies?.['expo-dev-client'] || !!pkg.devDependencies?.['expo-dev-client'];
}
