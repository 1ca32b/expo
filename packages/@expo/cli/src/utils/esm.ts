export async function tryRequireThenImport<TModule>(moduleId: string): Promise<TModule> {
  try {
    return require(moduleId);
  } catch (requireError: any) {
    let importESM;
    try {
      // eslint-disable-next-line no-new-func
      importESM = new Function('id', 'return import(id);');
    } catch {
      importESM = null;
    }

    if (requireError?.code === 'ERR_REQUIRE_ESM' && importESM) {
      return await importESM(moduleId);
    }

    throw requireError;
  }
}
