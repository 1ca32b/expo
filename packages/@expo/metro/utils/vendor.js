/**
 * Get the relative path of a file within a vendored package.
 * This uses the absolute path of the file, and splits it from the actual package name.
 */
function relativePackagePath(file, { packageName }) {
  const relativeFilePath = `${file.dir}/${file.base}`.split(packageName).pop();
  return `${packageName}${relativeFilePath}`;
}

module.exports = {
  relativePackagePath,
};
