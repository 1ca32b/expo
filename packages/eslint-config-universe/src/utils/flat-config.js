function wrapPluginMeta(plugin, packageName) {
  if (!plugin.meta) {
    plugin.meta = {
      name: packageName,
      version: require(`${packageName}/package.json`).version,
    };
  }

  return plugin;
}

function legacyPlugin(plugin) {
  return wrapPluginMeta(require(plugin), plugin);
}

module.exports = {
  legacyPlugin,
  wrapPluginMeta,
};
