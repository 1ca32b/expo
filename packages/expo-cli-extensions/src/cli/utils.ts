import { ExpoCliApplication } from './types';

export class SendMessageError extends Error {
  constructor(
    message: string,
    public app: ExpoCliApplication
  ) {
    super(message);
  }
}

export const getDeviceIdentifier = (app: ExpoCliApplication) => {
  // Use the deviceName + app ID as the device identifier
  return formatDeviceIdentifier(app.deviceName, app.appId);
};

export const formatDeviceIdentifier = (deviceName: string, applicationId: string) => {
  // Use the deviceName + app ID as the device identifier
  return `${deviceName} (${applicationId})`;
};
