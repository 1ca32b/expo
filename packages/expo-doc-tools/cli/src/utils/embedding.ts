let embeddingPipeline: any = null;

export async function embedText(text: string): Promise<number[]> {
  if (!embeddingPipeline) {
    const { pipeline } = await import('@xenova/transformers');
    embeddingPipeline = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');
  }
  // The pipeline returns [1, N, D] shape, so we flatten to [D]
  const output = await embeddingPipeline(text, {
    pooling: 'mean',
    normalize: true,
  });
  return Array.from(output.data);
}
