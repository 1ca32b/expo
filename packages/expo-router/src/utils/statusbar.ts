import Constants from 'expo-constants';
import { isEdgeToEdge } from 'react-native-is-edge-to-edge';

const hasViewControllerBasedStatusBarAppearance =
  process.env.EXPO_OS === 'ios' &&
  !!Constants.expoConfig?.ios?.infoPlist?.UIViewControllerBasedStatusBarAppearance;

export const canOverrideStatusBarBehavior =
  !isEdgeToEdge() && !hasViewControllerBasedStatusBarAppearance;
