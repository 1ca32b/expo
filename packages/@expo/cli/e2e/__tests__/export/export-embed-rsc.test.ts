/* eslint-env jest */
import JsonFile from '@expo/json-file';
import execa from 'execa';
import klawSync from 'klaw-sync';
import path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';

import { runExportSideEffects } from './export-side-effects';
import { bin, getRouterE2ERoot } from '../utils';
import { remove } from 'fs-extra';

runExportSideEffects();

describe('export embed for RSC iOS', () => {
  const projectRoot = getRouterE2ERoot();
  const outputName = 'dist-export-embed-rsc';
  const outputDir = path.join(projectRoot, outputName);

  beforeAll(
    async () => {
      await remove(outputDir);

      await execa(
        'node',
        [
          bin,
          'export:embed',
          //
          '--entry-file',
          path.join(projectRoot, './index.js'),
          //
          '--bundle-output',
          `./${outputName}/index.js`,
          '--assets-dest',
          outputName,
          '--platform',
          'ios',
          '--dev',
          'false',

          '--sourcemap-output',
          path.join(projectRoot, `./${outputName}/index.js.map`),

          '--sourcemap-sources-root',
          projectRoot,
        ],
        {
          cwd: projectRoot,
          env: {
            NODE_ENV: 'production',

            E2E_ROUTER_SRC: 'rsc',
            E2E_ROUTER_ASYNC: 'development',

            EXPO_USE_STATIC: 'single',
            E2E_ROUTER_JS_ENGINE: 'hermes',

            E2E_RSC_ENABLED: '1',
            E2E_CANARY_ENABLED: '1',
            EXPO_USE_METRO_REQUIRE: '1',
            TEST_SECRET_VALUE: 'test-secret',
          },
        }
      );
    },
    // Could take 45s depending on how fast the bundler resolves
    560 * 1000
  );

  it('has expected files', async () => {
    // List output files with sizes for snapshotting.
    // This is to make sure that any changes to the output are intentional.
    // Posix path formatting is used to make paths the same across OSes.
    const files = klawSync(outputDir)
      .map((entry) => {
        if (entry.path.includes('node_modules') || !entry.stats.isFile()) {
          return null;
        }
        return path.posix.relative(outputDir, entry.path);
      })
      .filter(Boolean);
    console.log(files);

    expect(
      await fs.promises.readFile(path.resolve(outputDir, '_flight/ios/index.txt'), 'utf8')
    ).toMatchSnapshot();

    // Ensure the standard files are included.
    expect(fs.existsSync(path.resolve(outputDir, 'index.js'))).toBe(true);
    expect(fs.existsSync(path.resolve(outputDir, 'assets'))).toBe(true);

    // Ensure the SSG RSC payload is included in the binary
    expect(fs.existsSync(path.resolve(outputDir, '_flight'))).toBe(true);

    // Ensure the secure server files are not included.
    expect(fs.existsSync(path.resolve(outputDir, 'client'))).toBe(false);
    expect(fs.existsSync(path.resolve(outputDir, 'server'))).toBe(false);

    // Check the server project location
    expect(fs.existsSync(path.resolve(projectRoot, '.expo/server/ios'))).toBe(true);
  });
});
