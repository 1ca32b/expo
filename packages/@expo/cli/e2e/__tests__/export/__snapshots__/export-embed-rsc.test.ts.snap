// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`export embed for RSC iOS has expected files 1`] = `
"1:I["apps/router-e2e/__e2e__/01-rsc/lib/react-native.tsx",[],"View"]
2:I["packages/expo-router/build/rsc/router/host.js",[],"Children"]
3:I["apps/router-e2e/__e2e__/01-rsc/lib/react-native.tsx",[],"Text"]
4:I["apps/router-e2e/__e2e__/01-rsc/lib/react-native.tsx",[],"Image"]
5:I["apps/router-e2e/assets/icon.png",[],""]
6:I["apps/router-e2e/__e2e__/01-rsc/components/counter.tsx",[],"Counter"]
0:{"layout":["$","$L1",null,{"style":{"flex":1,"padding":12},"testID":"layout-child-wrapper","children":[["$","$L2",null,{}],["$","$L1",null,{"testID":"layout-global-style","style":[{"width":100,"height":100},{"$$css":true,"_":"custom-global-style"}]}],["$","$L1",null,{"testID":"layout-module-style","style":[{"width":100,"height":100},"$undefined"]}]]}],"page":["$","$L1",null,{"style":{"flex":1,"padding":12},"testID":"child-wrapper","children":[["$","$L3",null,{"testID":"index-text","children":"Hello World"}],["$","$L3",null,{"testID":"index-path","children":"/"}],["$","$L3",null,{"testID":"index-query","children":""}],["$","$L3",null,{"children":["Platform: ","ios"]}],["$","$L3",null,{"testID":"secret-text","children":["Secret: ","test-secret"]}],["$","$L3",null,{"children":["Render: ",1726480496885]}],["$","$L4",null,{"testID":"main-image","source":"$5","style":{"width":100,"height":100}}],["$","$L6",null,{}]]}],"/SHOULD_SKIP":[["layout",[]]],"/LOCATION":["/",""]}
"
`;
